File: medium/1814_count_nice_pairs_in_an_array/solution.py
from collections import defaultdict
from typing import List


class Solution:
    def countNicePairs(self, nums: List[int]) -> int:
        pairsMap = defaultdict(int)

        for num in nums:
            numCopy = num
            numReversed = 0

            while numCopy:
                numReversed = numReversed * 10 + numCopy % 10
                numCopy //= 10

            pairsMap[num - numReversed] += 1

        MOD = 1000000007
        nicePairsCount = 0

        for count in pairsMap.values():
            nicePairsCount = (nicePairsCount + count * (count - 1) // 2) % MOD

        return nicePairsCount

File: medium/1814_count_nice_pairs_in_an_array/solution.cpp
#include <unordered_map>
#include <vector>

class Solution {
public:
  int countNicePairs(std::vector<int> &nums) {
    std::unordered_map<int, long long> pairsMap;

    for (const int &num : nums) {
      int numCopy = num;
      int numReversed = 0;

      while (numCopy) {
        numReversed = numReversed * 10 + numCopy % 10;
        numCopy /= 10;
      }

      ++pairsMap[num - numReversed];
    }

    const long long MOD = 1'000'000'007;
    long long nicePairsCount = 0;

    for (const auto &[_, count] : pairsMap)
      nicePairsCount = (nicePairsCount + count * (count - 1) / 2) % MOD;

    return static_cast<int>(nicePairsCount);
  }
};
